projects/js-packages/connection/components/disconnect-dialog/index.jsx

```diff
@@ -45,54 +69,211 @@ const DisconnectDialog = props => {
+						jetpackAnalytics.tracks.recordEvent(
+							'jetpack_disconnect_survey_error',
+							tracksSurveyData
+						);
+					}
+
+					setIsFeedbackProvided( true );
+					setIsSubmittingFeedback( false );
+				} );
 		},
-		[ setOpen ]
+		[ setIsSubmittingFeedback, setIsFeedbackProvided ]
 	);
 
 	/**
 	 * Disconnect - Triggered upon clicking the 'Disconnect' button.
 	 */
-	const disconnect = useCallback(
+	const handleDisconnect = useCallback(
 		e => {
 			e && e.preventDefault();
 
 			setDisconnectError( false );
 			setIsDisconnecting( true );
 
-			restApi
-				.disconnectSite()
-				.then( () => {
-					setIsDisconnecting( false );
+			// Detect the plugin context, where the plugin needs to be deactivated.
+			if ( context === 'plugins' ) {
+				// Use a callback function to handle deactivating the plugin.
+				// This should effectively short-circuit the disconnect flow by redirecting to deactivate the plugin.
+				if ( pluginScreenDisconnectCallback ) {
+					pluginScreenDisconnectCallback( e );
+				}
+				// Do not disconnect if context is the plugin screen, the plugin deactivation routine will handle disconnection.
+				return;
+			}
 
-					setIsDisconnected( true );
-				} )
-				.catch( error => {
-					setIsDisconnecting( false );
-					setDisconnectError( error );
+			// Default to making the disconnect API call here.
+			_disconnect();
+		},
+		[ setDisconnectError, setIsDisconnecting, pluginScreenDisconnectCallback, context, _disconnect ]
+	);
 
-					if ( onError ) {
-						onError( error );
-					}
-				} );
+	/**
+	 * Do we have the necessary data to be able to submit a survey?
+	 * Need to have the ID of the connected user and the ID of the connected site.
+	 */
+	const canProvideFeedback = useCallback( () => {
+		return connectedUser.ID && connectedSiteId;
+	}, [ connectedUser, connectedSiteId ] );
+
+	/**
+	 * Submit Survey - triggered by clicking on the "Submit Feedback" button.
+	 * Assembles the survey response.
+	 */
+	const handleSubmitSurvey = useCallback(
+		( surveyAnswerId, surveyAnswerText, e ) => {
+			e && e.preventDefault();
+
+			// We do not have the information needed to record the response.
+			// return early and move to the last step in the flow anyway.
+			if ( ! canProvideFeedback() ) {
+				setIsFeedbackProvided( true );
+				return;
+			}
+
+			// Format the survey data for submission.
+			const surveyData = {
+				site_id: connectedSiteId,
+				user_id: connectedUser.ID,
+				survey_id: 'jetpack-plugin-disconnect',
+				survey_responses: {
+					'why-cancel': {
+						response: surveyAnswerId,
+						text: surveyAnswerText ? surveyAnswerText : null,
+					},
+				},
+			};
+
+			// Additional data for analytics to see where disconnections happened from.
+			const tracksSurveyData = Object.assign( {}, defaultTracksArgs, {
+				disconnect_reason: surveyAnswerId,
+			} );
+
+			if ( surveyAnswerText ) {
+				surveyData.disconnect_reason_text = surveyAnswerText;
```

Comment:
Do we need this if it's already being passed in the `text` field within `survey_responses`?
When submitting the survey with a custom response I'm seeing the entered reason being passed in both fields.
<img width="603" alt="Screen Shot 2021-11-17 at 11 10 44 AM" src="https://user-images.githubusercontent.com/690843/142272597-b8e496e0-45a5-4c5b-949a-877c9d36a60a.png">

Reply:
Good catch! This was left-over from when we were collecting the survey with Tracks and got missed when I switched to using the new endpoint.